### PR TITLE
This migrates the `publish.NewDefault` method to functional options.

### PR DESCRIPTION
**Pull request by mattmoor**

### CODE REVIEW

File: pkg/ko/publish/options.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// Set client on image using provided transport, and the default authenticator
+func (i *defaultOpener) setTransport(t http.RoundTripper) error {
+	i.t = t
+	return nil
+}
+
+// Set client on image using provided authenticator, and the default transport
+func (i *defaultOpener) setAuth(auth authn.Authenticator) error {
+	i.auth = auth
+	return nil
```
Comment (ImJasonH) [inline pkg/ko/publish/options.go]: These don't need to `return nil`, I think it's simpler to just do that above in the returned func.

Comment (ImJasonH) [reply pkg/ko/publish/options.go]: You probably don't even need `setAuth`? Just `i.auth = auth; return nil` above in `WithAuth`

Comment (mattmoor) [reply pkg/ko/publish/options.go]: Should be RFAL

File: pkg/ko/publish/options.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		if err != nil {
+			return err
+		}
+		i.setAuth(auth)
+		return nil
+	}
+}
+
+// Set client on image using provided transport, and the default authenticator
+func (i *defaultOpener) setTransport(t http.RoundTripper) {
```
Comment (ImJasonH) [inline pkg/ko/publish/options.go]: I think even `setTransport` and `setAuth` are pretty useless, and could just be removed, but I'll leave that up to you.

**Pull request merged**